test/unit-test/mqtt_agent_command_functions_utest.c

```diff
@@ -73,3 +178,338 @@ int suiteTearDown( int numFailures )
+/**
+ * @brief Test MQTTAgentCommand_Publish() success case with QoS1.
+ */
+void test_MQTTAgentCommand_Publish_QoS1_success( void )
+{
+    MQTTAgentContext_t mqttAgentContext = { 0 };
+    MQTTPublishInfo_t publishInfo = { 0 };
+    MQTTAgentCommandFuncReturns_t returnFlags = { 0 };
+    MQTTStatus_t mqttStatus;
+
+    /* Initializing QOS. */
+    publishInfo.qos = MQTTQoS1;
+
+    MQTT_GetPacketId_ExpectAndReturn( &( mqttAgentContext.mqttContext ), 1 );
+    MQTT_Publish_ExpectAndReturn( &( mqttAgentContext.mqttContext ), &publishInfo, 1, MQTTSuccess );
+
+    mqttStatus = MQTTAgentCommand_Publish( &mqttAgentContext, &publishInfo, &returnFlags );
+
+    TEST_ASSERT_EQUAL( MQTTSuccess, mqttStatus );
+    /* Ensure that returnFlags are set as intended. */
+    TEST_ASSERT_EQUAL( 1, returnFlags.packetId );
+    TEST_ASSERT_TRUE( returnFlags.runProcessLoop );
+    TEST_ASSERT_TRUE( returnFlags.addAcknowledgment );
+    TEST_ASSERT_FALSE( returnFlags.endLoop );
+}
+
+/**
+ * @brief Test MQTTAgentCommand_Publish() failure case with QoS0.
+ */
+void test_MQTTAgentCommand_Publish_QoS0_failure( void )
+{
+    MQTTAgentContext_t mqttAgentContext = { 0 };
+    MQTTPublishInfo_t publishInfo = { 0 };
+    MQTTAgentCommandFuncReturns_t returnFlags = { 0 };
+    MQTTStatus_t mqttStatus;
+
+    /* Initializing QOS. */
+    publishInfo.qos = MQTTQoS0;
+
+    MQTT_Publish_ExpectAndReturn( &( mqttAgentContext.mqttContext ), &publishInfo, 0, MQTTSendFailed );
+
+    mqttStatus = MQTTAgentCommand_Publish( &mqttAgentContext, &publishInfo, &returnFlags );
+
+    TEST_ASSERT_EQUAL( MQTTSendFailed, mqttStatus );
+    /* Ensure that returnFlags are set as intended. */
+    TEST_ASSERT_EQUAL( 0, returnFlags.packetId );
+    TEST_ASSERT_TRUE( returnFlags.runProcessLoop );
+    TEST_ASSERT_FALSE( returnFlags.addAcknowledgment );
+    TEST_ASSERT_FALSE( returnFlags.endLoop );
+}
+
+/**
+ * @brief Test MQTTAgentCommand_Publish() failure case with QoS1.
+ */
+void test_MQTTAgentCommand_Publish_QoS1_failure( void )
+{
+    MQTTAgentContext_t mqttAgentContext = { 0 };
+    MQTTPublishInfo_t publishInfo = { 0 };
+    MQTTAgentCommandFuncReturns_t returnFlags = { 0 };
+    MQTTStatus_t mqttStatus;
+
+    /* Initializing QOS. */
+    publishInfo.qos = MQTTQoS1;
+
+    MQTT_GetPacketId_ExpectAndReturn( &( mqttAgentContext.mqttContext ), 1 );
+    MQTT_Publish_ExpectAndReturn( &( mqttAgentContext.mqttContext ), &publishInfo, 1, MQTTSendFailed );
+
+    mqttStatus = MQTTAgentCommand_Publish( &mqttAgentContext, &publishInfo, &returnFlags );
+
+    TEST_ASSERT_EQUAL( MQTTSendFailed, mqttStatus );
+    /* Ensure that returnFlags are set as intended. */
+    TEST_ASSERT_EQUAL( 1, returnFlags.packetId );
+    TEST_ASSERT_TRUE( returnFlags.runProcessLoop );
+    TEST_ASSERT_FALSE( returnFlags.addAcknowledgment );
+    TEST_ASSERT_FALSE( returnFlags.endLoop );
+}
+
+/**
+ * @brief Test that MQTTAgentCommand_Subscribe() works as intended.
+ */
+void test_MQTTAgentCommand_Subscribe( void )
```

Comment:
We have publish success & failure unit tests, why are the subscribe/unsubscribe tests only covering success cases?

Reply:
The tests for mqttagent_subscribe and mqttagent_unsubscribe calls the mqtt_subscribe and mqtt_unsubscribe mock functions , so failure cases depends only on the values they return and these mock functions failure paths has been already tested in coreMQTT unit tests. There is no specific path that mqttagent functions needs to take on failure. So if you recommend I can include those test cases too.